### DISCUSSION TITLE
half precision support at gemm_complex.h

### DISCUSSION THREAD

**Discussion by YoniSp** [NVIDIA/cutlass#501] [Q&A]
Hi,
I would like to know if gemm_complex.h support complex multiplication while the input is complex<bfloat16_t> and output is complex<float>.
If there is support for this configuration I would like to know if there is any example for this constellation?

Regards

Yehonatan

**Comment by hwu36** [comment]
It is not going to work out of the box.  You need to fill some holes to make it work.

**Comment by YoniSp** [comment]
Hi,
Thanks for the reply.
I am new with this library, can you be more specific about which holes should I cover and how to fix them?

**Comment by hwu36** [comment]
Do you have to work on `complex<>` or you can separate real and image part before calling the kernel?   The latter is easier for tensor cores which is like https://github.com/NVIDIA/cutlass/blob/master/examples/10_planar_complex/planar_complex.cu because you don't need to unpack and pack the data after loading them from the shared memory to the registers.

**Comment by YoniSp** [comment]
Hi,
Currently data layout is nit planner but complex<> in that case which modifications should I do in order to make it work with cutlass gemm_complex.h.

**Comment by hwu36** [comment]
You need to add data packing unpacking similar to https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/warp/mma_complex_tensor_op_fast_f32.h#L100

You need to add partial specialization for fp16 in https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/warp/default_mma_complex_tensor_op.h

You need to add partial specialization for fp16 in https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/gemm/threadblock/default_multistage_mma_complex_core_sm80.h

There maybe more.
